server/src/main/java/io/druid/server/http/DatasourcesResource.java

```diff
@@ -496,40 +493,32 @@ public Response getSegmentDataSourceTiers(
     return Response.ok(retVal).build();
   }
 
-  private DruidDataSource getDataSource(final String dataSourceName)
+  private ImmutableDruidDataSource getDataSource(final String dataSourceName)
   {
-    Iterable<DruidDataSource> dataSources =
-        Iterables.concat(
-            Iterables.transform(
-                serverInventoryView.getInventory(),
-                (DruidServer input) -> input.getDataSource(dataSourceName)
-            )
-        );
-
-    List<DruidDataSource> validDataSources = Lists.newArrayList();
-    for (DruidDataSource dataSource : dataSources) {
-      if (dataSource != null) {
-        validDataSources.add(dataSource);
-      }
-    }
-    if (validDataSources.isEmpty()) {
+    List<ImmutableDruidDataSource> dataSources = serverInventoryView
+        .getInventory()
+        .stream()
+        .map(server -> server.getDataSource(dataSourceName))
+        .filter(Objects::nonNull)
+        .map(DruidDataSource::toImmutableDruidDataSource)
+        .collect(Collectors.toList());
+
+    if (dataSources.isEmpty()) {
       return null;
     }
 
     Map<String, DataSegment> segmentMap = Maps.newHashMap();
-    for (DruidDataSource dataSource : validDataSources) {
-      if (dataSource != null) {
-        Iterable<DataSegment> segments = dataSource.getSegments();
-        for (DataSegment segment : segments) {
-          segmentMap.put(segment.getIdentifier(), segment);
-        }
+    for (ImmutableDruidDataSource dataSource : dataSources) {
+      Iterable<DataSegment> segments = dataSource.getSegments();
+      for (DataSegment segment : segments) {
+        segmentMap.put(segment.getIdentifier(), segment);
       }
     }
 
     return new DruidDataSource(
```

Comment:
This is a strange expression, why ImmutableDruidDataSource couldn't be constructed directly?

Reply:
Ah, it was somewhat complicated before, but now it's fine. Changed to directly construct ImmutableDruidDataSource.
Note that ImmutableMap.Builder cannot be used for segmentMap because the same segment identifier can be inserted multiple times. I'm not sure this is the valid behavior and want to keep it as it is.